lib/collection/src/shards/shard_holder/mod.rs

```diff
@@ -581,6 +593,48 @@ impl ShardHolder {
         Ok(res)
     }
 
+    /// Common routing logic for reads and writes when using ShardKeyWithFallback
+    ///
+    /// If at least one of target shards is Active, use target shard. If not, redirect to fallback shard
+    pub fn route_with_fallback<'a>(
+        &self,
+        key: &'a ShardKeyWithFallback,
+    ) -> CollectionResult<(HashSet<ShardId>, &'a ShardKey)> {
+        let shard_key_to_ids_mapping = self.get_shard_key_to_ids_mapping();
+
+        let target_shard_ids = shard_key_to_ids_mapping.get(&key.target);
+        let fallback_shard_ids = shard_key_to_ids_mapping.get(&key.fallback);
+
+        if let Some(target_shard_ids) = target_shard_ids {
+            let replicas = target_shard_ids
+                .iter()
+                .filter_map(|shard_id| self.shards.get(shard_id))
+                .collect::<Vec<_>>();
+
+            let target_any_active = replicas
+                .iter()
+                .any(|replica_set| !replica_set.active_shards().is_empty());
```

Comment:
`any` -> `all` ?

My intuition: `replica_set` is a set of replicas per shard-id. We do expect that target shard has only one `replica_set`, but if not, we would want all shard_ids to have at least one active shard